core/src/main/java/com/linecorp/armeria/client/endpoint/PropertiesFileWatcherRegistry.java

```diff
@@ -0,0 +1,173 @@
+        }
+    }
+
+    /**
+     * Registers a resourceUrl of the properties file to be watched.
+     * @param resourceUrl url to be watched
+     * @param reloader callback to be called on a file change event
+     */
+    void register(URL resourceUrl, Runnable reloader) {
+        final File file = new File(resourceUrl.getFile());
+        final Path path = file.getParentFile().toPath();
+
+        synchronized (ctxRegistry) {
+            checkArgument(!ctxRegistry.containsKey(resourceUrl.getFile()),
+                          "file is already watched: %s", resourceUrl.getFile());
+            try {
+                final WatchKey key = path.register(watchService,
+                                                   ENTRY_CREATE,
+                                                   ENTRY_MODIFY,
+                                                   ENTRY_DELETE,
+                                                   OVERFLOW);
+                ctxRegistry.put(resourceUrl.getFile(), new PropertiesFileWatcherContext(key, reloader));
+            } catch (IOException e) {
+                throw new IllegalArgumentException("failed to watch file " + resourceUrl.getFile(), e);
+            }
+        }
+
+        startFutureIfPossible();
+    }
+
+    /**
+     * Stops watching a properties file corresponding to the resourceUrl.
+     * @param resourceUrl url to stop watching
+     */
+    void deregister(@Nonnull URL resourceUrl) {
+        final PropertiesFileWatcherContext context = ctxRegistry.remove(resourceUrl.getFile());
+        context.key.cancel();
+        stopFutureIfPossible();
+    }
+
+    private synchronized void startFutureIfPossible() {
+        if (!isRunning() && !ctxRegistry.isEmpty()) {
+            future = CompletableFuture.runAsync(new PropertiesFileWatcherRunnable(), eventLoop);
+        }
+    }
+
+    private synchronized void stopFutureIfPossible() {
+        Preconditions.checkState(future != null, "tried to stop null future");
+        if (isRunning() && ctxRegistry.isEmpty()) {
+            future.cancel(true);
+        }
+    }
+
+    private boolean isRunning() {
+        return future != null && !future.isDone();
+    }
+
+    private class PropertiesFileWatcherRunnable implements Runnable {
+        @Override
+        public void run() {
+            try {
+                WatchKey key;
+                while ((key = watchService.take()) != null) {
+                    for (WatchEvent<?> event : key.pollEvents()) {
+                        @SuppressWarnings("unchecked")
+                        final Path watchedPath = ((Path) key.watchable())
+                                .resolve(((WatchEvent<Path>) event).context());
+                        final String watchedPathFile = watchedPath.toFile().getAbsolutePath();
+
+                        if (event.kind().equals(ENTRY_MODIFY) || event.kind().equals(ENTRY_CREATE)) {
+                            if (ctxRegistry.containsKey(watchedPathFile)) {
```

Comment:
There are certain environments where this simple check doesn't actually work from what I've seen. The file might be a symlink, or the directory, not the file might change. I've had better luck iterating, and comparing the real path using `startsWith`

https://github.com/curioswitch/curiostack/blob/master/common/server/framework/src/main/java/org/curioswitch/common/server/framework/files/FileWatcher.java#L140